Add Configuration

mrocklin opened the pull request:
This adds a basic config file and uses that configuration throughout the
dask-yarn codebase.

This also switches the convention away from providing memory size in
megabytes and instead using strings with units.  We provide informative
error messages to guide users to correct behavior.

cc @jcrist for review.  In particular I want to check in and make sure that you're ok with the memory handling conventions.  They're different from skein.  We can walk back from this a bit if you like.

mrocklin commented:
@mariusvniekerk

quasiben commented:
Note, I'll probably only get to writing tests for this later this today or possibly tomorrow ?

quasiben commented:
Nevermind, test was easy.  Thought I had to spin up yarn to get this going.  It did reveal a bit of clunkiness.  Without setting `autodetect=False` the following will result in nn/rm (most likely) being set to `localhost`

```
YARNCluster(nn="pi", nn_port=31415, rm="e", rm_port=27182..)
```

Perhaps instead I should check if autodetect is false or any values move away from there defaults

mariusvniekerk commented:
Cool

quasiben commented:
@mariusvniekerk have you had a chance to test?  Ideally i can merge in the next hour or so

mrocklin committed:
rename make_remote_spec to make_specification

mrocklin committed:
Add configuration

This adds a basic config file and uses that configuration throughout the
dask-yarn codebase.

This also switches the convention away from providing memory size in
megabytes and instead using strings with units.  We provide informative
error messages to guide users to correct behavior.

mrocklin committed:
add requirements.txt

mrocklin committed:
add test and error around passing keywords to constructor

jcrist commented on MANIFEST.in:
```diff
@@ -0,0 +1,12 @@
+recursive-exclude * __pycache__
+recursive-exclude * *.py[co]
+include versioneer.py
+include dask_yarn/_version.py
```
versioneer isn't currently used.

mrocklin replied:
Fixed

jcrist commented on dask_yarn/core.py:
```diff
@@ -29,38 +52,81 @@ def make_remote_spec(environment, name='dask', queue='default', tags=None,
-    worker_memory : int, optional
-        The ammount of memory in MB to allocate per worker.
+    worker_memory : str, optional
+        The ammount of memory to allocate per worker like '2GB' or '4096 MiB'.
```
Could we support both strings and int (in MB)?

mrocklin replied:
We *could*, though I think it's a strange default.  Of course, it could be argued that it's a default choice that's made for us by Yarn.  Current convention in Dask is that if the user provides a number then that number corresponds to the number of bytes, not megabytes.

It's pretty obvious today what the user wanted, and we can automatically convert any number less than 1e6 to megabytes if that would be nicer.  I somewhat prefer the current convention because it is fully explicit, but I can see how it might be unpleasant to a user familiar with YARN conventions.  Happy to make the change.

jcrist replied:
It's easier to add later than to remove. Fine as is.

jcrist commented on setup.py:
```diff
@@ -1,11 +1,15 @@
 from setuptools import setup
 
+with open('requirements.txt') as f:
+    install_requires = f.read().strip().split('\n')
```
I personally prefer to keep it all in the setup.py file, but this isn't a strong preference.

mrocklin replied:
Most projects today seem to keep things in requirements.txt .  I'm inclined to stay with that convention.

I also like this personally because it's easy to look up what a package requires quickly in a data file rather than looking through a setup.py file.  Although in this case the setup.py is simple enough that this isn't a very solid argument.  The "common convention" one though I think does stand.

jcrist commented on dask_yarn/core.py:
```diff
@@ -29,38 +52,81 @@ def make_remote_spec(environment, name='dask', queue='default', tags=None,
+        msg = ("You must provide a path to a redeployable environment for the "
+               "workers.\n"
+               "This is commonly achieved through conda-pack.\n\n"
+               "See https://dask-yarn.readthedocs.org/en/latest/environments.html "
```
Does this docs page exist?

mrocklin replied:
Not yet!  I was going to play with documentation next

mrocklin committed:
remove versioneer file references from MANIFEST
